### ISSUE TITLE
ACL not respected for shared node in share recipient's tailnet

### ISSUE THREAD

**Issue by e-nomem** [tailscale/tailscale#4302]
### Scenario
A family member with a separate tailscale account shared a machine running pihole with my account. Within my account, I have an ACL rule allowing my machines to use the pihole for DNS, and have configured it as the DNS override for the account, but do not allow any machine to access the admin web UI on port 80. Despite this, I am still able to access port 80/the web UI from my machines.

I subsequently removed all ACL rules allowing any access to the pihole whatsoever, but my DNS configuration did not stop working, and all DNS queries were still being successfully answered by the pihole (I ran `dig A google.com @<pihole_tailscale_address>`)

### Steps to reproduce

1. In tailnet A, add a pihole machine (here with IP address `100.xxx.xxx.xxx`)
  a. This probably works with any shared machine, but pihole is what I experimented with
3. Share this machine with tailnet B
4. For tailnet A, use the following ACL:
```hujson
{
  "hosts": {
    "pihole": "100.xxx.xxx.xxx"
  },
  "acls": [
    // Allow access to everything
    {"action": "accept", "src": ["*"], "dst": ["*:*"]}
  ],
  "tests": [
    {
      "src": "user.a.email@gmail.com",
      "accept": [
        // Everything is accessible within tailnet A
        "pihole:53",
        "pihole:80"
      ]
  ]
}
```
5. For tailnet B, use the following ACL:
```hujson
{
  "hosts": {
    "pihole": "100.xxx.xxx.xxx"
  },
  "acls": [
    // Allow access to nothing
  ],
  "tests": [
    {
      "src": "user.b.email@gmail.com",
      "deny": [
        // Pihole not accessible within tailnet B
        "pihole:53",
        "pihole:80"
      ]
    }
  ]
}
```
6. Connect to tailscale on a device within tailnet B
7. Access the pihole machine (e.g. `dig A google.com @100.xxx.xxx.xxx` or open `http://100.xxx.xxx.xxx/admin` in the browser)

### OS

macOS, Windows, Android

### OS version

Android 12, macOS 12.3, Windows 10

### Tailscale version

1.22.1 (macOS), 1.22.2 (windows), 1.23.148 (android)

### Bug report

BUG-d842ee1acc7d4800e27a74bf05804a67721f817652139edeade3a84dac1ef4d1-20220328060605Z-bcef9869e698cf6f

**Comment by danderson** [comment]
This is working as intended (although your report indicates that we need to improve our documentation on this front).

ACLs for shared machines can only be defined by the machine's owner, in tailnet A in your example. ACLs in Tailscale are all enforced at the destination only, so if you define a restriction on who can connect to machine X, it's machine X and only machine X that receives a packet filter with those rules. Since tailnet B doesn't own the shared machine, it can't have an opinion on what packet filter to send it.

In theory, we could enforce the restrictions from tailnet B's ACLs on the source machines, which do belong to tailnet B... but enforcement-at-source is fundamentally untrustable, because ACLs are about protecting a target from access by unauthorized sources, and an unauthorized source could just run a modified tailscaled that ignores the outbound packet filter.

So, to achieve the lockdown you're looking for, the owner of tailnet A needs to configure ACLs that restrict what shared-to users can do with their machine. Usually, this would be done with a rule that uses `autogroup:shared` as the `src`, for example:

```json
{
  // Allow all members of tailnet A to access all of Tailnet A
  "src": ["autogroup:members"],
  "dst": ["*:*"],
},
{
  // Allow people who accepted a share of pihole to access port 53, and only port 53.
  "src": ["autogroup:shared"],
  "dst": ["pihole:53"],
}
```

**Comment by e-nomem** [comment]
That makes sense. A documentation update is probably sufficient to address this then. In https://tailscale.com/kb/1084/sharing/#sharing--access-control-lists-acls it states:
```
Sharing respects the ACLs of both networks. If either network disallows connections to/from a device, no connections can be established. If you’re having trouble connecting to a shared device, review your ACL file for rules that might be blocking access.
```
which gives the impression that I should be able to further restrict the permissions beyond what the owner allowed.

**Comment by danderson** [comment]
Oh yeah, that wording's quite confusing, I can see how that led you astray. Thanks for reporting it, we'll get it fixed!

**Comment by e-nomem** [comment]
One final though, perhaps as a feature request? Given that the ACLs in tailnet A are opaque from tailnet B, maybe ACL tests  referencing external/shared machines should be rejected as 'unenforceable' or 'unverifiable'? Seeing my passing deny tests gives me false confidence that something is being restricted when it actually isn't.

In any case, thanks for the quick response. I'm loving tailscale and I've worked with my whole family to move to it over the weekend 😁

**Comment by danderson** [comment]
Yeah, good point, that test should fail in some way if it's not going to be enforced in practice. I think the way the tests are defined right now, they don't care what kind of node is being tested, just that "if I was asked to generate a filter for the `pihole` node, then the test would pass". It never learns that this filter is irrelevant.

**Comment by mayakacz** [comment]
cc @WalterHub

**Comment by Biepa** [comment]
> Given that the ACLs in tailnet A are opaque from tailnet B, maybe ACL tests referencing external/shared machines should be rejected as 'unenforceable' or 'unverifiable'?

Yes please do that. Came by the same issue at it's quite confusing.
